File: 139.单词拆分.cpp
/*
 * @lc app=leetcode.cn id=139 lang=cpp
 *
 * [139] 单词拆分
 */

// @lc code=start
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int size = s.size();
        auto dp = vector<bool>(size + 1, false);
        dp[0] = true;
        for (int i = 0; i <= size; i++) {
            if (!dp[i]) continue;
            for (string &item : wordDict) {
                int len = item.size();
                if (i + len > size) continue;
                // 确认是否相同
                dp[i + len] =  dp[i + len] || isSame(s, i, i - 1 + len, item) ; 
            }
        }
        return dp[size];
    }

    bool isSame(string &s, int startPos, int endPos, string &compare) {
        if (endPos >= s.size()) return false;
        
        for (int i = startPos, j = 0; i <= endPos; i++, j++) {
            if (s[i] != compare[j]) return false;
        }
        return true;
    }
};
// @lc code=end

File: 139.单词拆分.swift
// swiftlint:disable identifier_name trailing_newline
/*
 * @lc app=leetcode.cn id=139 lang=swift
 *
 * [139] 单词拆分
 */

// @lc code=start
class Solution {
    func wordBreak(_ s: String, _ wordDict: [String]) -> Bool {
        let size = s.count
        var dp = Array(repeating: false, count: size + 1)
        dp[0] = true
        for i in 0...size {
            if !dp[i] { continue }
            for word in wordDict {
                let len = word.count
                if i + len > size { continue }
                let startIndex = s.index(s.startIndex, offsetBy: i)
                let endIndex = s.index(s.startIndex, offsetBy: i + len - 1)

                if s[startIndex...endIndex] == word { dp[i+len] = dp[i+len] || true }
            }
        }
        return dp[size]
    }
}
// @lc code=end

